### PR TITLE
Add EtcdFix lib

### PR DESCRIPTION
**Pull request by ydye**

### CODE REVIEW

File: kubernetes-deployment/maintainlib/etcdfix.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        ip_peer = "{0}=http://{1}:2380".format(new_etcd_id, peer_url)
+        etcd_cluster_ips_peer = etcd_cluster_ips_peer + separated + ip_peer
+
+        self.logger.debug("New etcd-initial-cluster: {0}".format(etcd_cluster_ips_peer))
+
+        return etcd_cluster_ips_peer
+
+
+
+    def run(self):
```
Comment (fanyangCS) [inline kubernetes-deployment/maintainlib/etcdfix.py]: do you want to add a repair-infra-node.sh to demonstrate how to use the class?

File: kubernetes-deployment/maintainlib/etcdfix.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+
+
+class etcdfix:
+
+    """
+    A class to reconfiguration etcd. Fix the issue when etcd's data is corrupted
+    """
+
+    def __init__(self, cluster_config, node_config, clean):
```
Comment (fanyangCS) [inline kubernetes-deployment/maintainlib/etcdfix.py]: >cluster_config, node_config [](start = 23, length = 27)

can you add ut to express you expectation of the input?
cluster_config['mastermachinelist']
cluster_config['clusterinfo']['etcd_cluster_ips_peer'] =         self.cluster_config['clusterinfo']['etcd-initial-cluster-state']

bad_node_config['etcdid']
bad_node_config['hostip']

File: kubernetes-deployment/maintainlib/etcdfix.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+        self.logger.debug("Bad node information: {0}".format(str(bad_node_config)))
+
+        self.stop_bad_etcd_server(bad_node_config)
+
+        while True:
+
+            good_node_config = self.get_etcd_leader_node()
+
+            if good_node_config['nodename'] != self.bad_node_config['nodename']:
```
Comment (fanyangCS) [inline kubernetes-deployment/maintainlib/etcdfix.py]: >bad_node_config [](start = 52, length = 15)

what if bad node and good node don't overlap? you should add a unit test

File: kubernetes-deployment/maintainlib/etcdfix.py (hunk 1 above)
Comment (ydye) [reply kubernetes-deployment/maintainlib/etcdfix.py]: Later

File: kubernetes-deployment/maintainlib/etcdfix.py (hunk 3 above)
Comment (ydye) [reply kubernetes-deployment/maintainlib/etcdfix.py]: I have no idea about what to test. Can you give me an example?

File: kubernetes-deployment/maintainlib/etcdfix.py (hunk 2 above)
Comment (ydye) [reply kubernetes-deployment/maintainlib/etcdfix.py]: Why should we add ut to test the correctness of the cluster configuration? This function is designed object initialization. I don't think we should add ut here.

Comment (fanyangCS) [reply kubernetes-deployment/maintainlib/etcdfix.py]: sorry, it is more than just ut. you should handle the format incorrectness in your code. Currently, your code implicitly assumes cluster_config has some fields available. e.g., ['mastermachinelist'], what if it doesn't have this field ? what is the expected behavior of your code?  and you should add ut to explicitly test your code's behavior under such ill defined input and meanwhile let code readers understand your purpose in the UT.

---
In reply to: [172112809](https://github.com/Microsoft/pai/pull/345#discussion_r172112809) [](ancestors = 172112809)

Comment (ydye) [reply kubernetes-deployment/maintainlib/etcdfix.py]: Let me try to figure out how to add ut. Close this PR currently.

Comment (ydye) [reply kubernetes-deployment/maintainlib/etcdfix.py]: Reopen this PR, and developing it in the status of Openning.

File: kubernetes-deployment/maintainlib/etcdfix.py [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+
+        bad_node_config = self.bad_node_config
+
+        self.logger.debug("Bad node information: {0}".format(str(bad_node_config)))
+
+        self.stop_bad_etcd_server(bad_node_config)
+
+        while True:
+
+            good_node_config = self.get_etcd_leader_node()
```
Comment (fanyangCS) [inline kubernetes-deployment/maintainlib/etcdfix.py]: >get_etcd_leader_node [](start = 36, length = 20)

what if it return none?

File: kubernetes-deployment/maintainlib/etcdfix.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+            return
+
+        self.logger.info("Successfully restarting bad etcd server on node {0}".format(bad_node_config["nodename"]))
+
+        if self.clean_flag:
+            self.delete_packege(bad_node_config)
+
+
+
+    def get_etcd_leader_node(self):
```
Comment (fanyangCS) [inline kubernetes-deployment/maintainlib/etcdfix.py]: >get_etcd_leader_node [](start = 8, length = 20)

I suggest you move etcd related operation to common lib

File: kubernetes-deployment/maintainlib/etcdfix.py (hunk 3 above)
Comment (fanyangCS) [reply kubernetes-deployment/maintainlib/etcdfix.py]: I see. I suggest to add a comment: "waiting for the bad node to demote from leader"

---
In reply to: [172111494](https://github.com/Microsoft/pai/pull/345#discussion_r172111494) [](ancestors = 172111494)

File: kubernetes-deployment/maintainlib/etcdfix.py (hunk 4 above)
Comment (ydye) [reply kubernetes-deployment/maintainlib/etcdfix.py]: return false

File: kubernetes-deployment/maintainlib/etcdfix.py (hunk 2 above)
Comment (ydye) [reply kubernetes-deployment/maintainlib/etcdfix.py]: Add validation class

File: kubernetes-deployment/maintainlib/etcdfix.py (hunk 5 above)
Comment (ydye) [reply kubernetes-deployment/maintainlib/etcdfix.py]: Better not. I hope one lib's modify won't have an influence on others.

File: kubernetes-deployment/maintainlib/etcdfix.py (hunk 3 above)
Comment (ydye) [reply kubernetes-deployment/maintainlib/etcdfix.py]: Done

**Pull request merged**